src/mscorlib/src/System/IO/MemoryStream.cs

```diff
@@ -402,31 +402,42 @@ public override Task<int> ReadAsync(Byte[] buffer, int offset, int count, Cancel
             return _buffer[_position++];
         }
 
+        public override void CopyTo(Stream destination, int bufferSize)
+        {
+            // Since we did not originally override this method, validate the arguments
+            // the same way Stream does for back-compat.
+            ValidateCopyToArguments(destination, bufferSize);
 
-        public override Task CopyToAsync(Stream destination, Int32 bufferSize, CancellationToken cancellationToken) {
-
-            // This implementation offers beter performance compared to the base class version.
-
-            // The parameter checks must be in sync with the base version:
-            if (destination == null)
-                throw new ArgumentNullException("destination");
+            // If we have been inherited into a subclass, the following implementation could be incorrect
+            // since it does not call through to Read() or Write() which a subclass might have overriden.  
+            // To be safe we will only use this implementation in cases where we know it is safe to do so,
+            // and delegate to our base class (which will call into Read/Write) when we are not sure.
+            if (GetType() != typeof(MemoryStream))
+            {
+                base.CopyTo(destination, bufferSize);
+                return;
+            }
             
-            if (bufferSize <= 0)
-                throw new ArgumentOutOfRangeException("bufferSize", Environment.GetResourceString("ArgumentOutOfRange_NeedPosNum"));
+            int originalPosition = _position;
+            int remaining = _length - originalPosition;
```

Comment:
@jamesqo I see it in a similar way as @stephentoub. Using the `InternalEmulateRead` is actually more readable to me, since its name clearly expresses what it does and I don't have to read three lines of differences and assignments and figure it out myself. 
I would also like to have the same piece of code in the sync and async methods.
Having said that, if you feel strongly about your current way, I would not block the PR, but if would be great to make it the same in the async method in this PR and add a comment that the goal of this code is to emulate the read till the end of the stream or something like that.